README.md
# lazywf

The laziest web framework ever.

[![CircleCI](https://circleci.com/gh/cleberzavadniak/lazywf/tree/master.svg?style=svg)](https://circleci.com/gh/cleberzavadniak/lazywf/tree/master)

## How to use

1- In your project root, create a `server.py` script this
way:

```python
#!env python3

from lazywf import TheLaziestWebFrameworkEVER


class Lazy(TheLaziestWebFrameworkEVER):
    pass

Lazy().run()
```


2- Then, create a `models.yaml` like this:

```yaml
products:  # Your model name
    constraints:  # Some constrains you'd like to enforce
        keys: [sku]
        unique: [sku]
    validations:  # Describe some validations for your data
        sku:
            type: string
            required: true
            minlength: 16
            maxlength: 18
        name:
            type: string
            required: true
            maxlength: 64
        price:
            type: float
            required: true
        old_price:
            type: float
            required: false
            nullable: true
            default: null
```

Data validation is done using the excelent
[Cerberus](https://github.com/pyeve/cerberus)
project.

3- Export your `DATABASE_URL` environment variable:

```bash
$ export DATABASE_URL="sqlite:///database.sqlite"
```

4- Now you can run your server:

```bash
$ python3 server.py
```

5- Test it using
[httpie](https://github.com/jakubroztocil/httpie):

```bash
$ http --json POST "http://localhost:8080/api/products/" sku=12345678901234567 name=TestProduct price:=19.90
$ http "http://localhost:8080/api/products/"
$ http --json PATCH "http://localhost:8080/api/products/12345678901234567" old_price:=19.90 price:=29.90
$ http "http://localhost:8080/api/products/"
$ http DELETE "http://localhost:8080/api/products/1"
$ http "http://localhost:8080/api/products/"
```


## Why I started this project

I really enjoy working with **Django**. It's a very nice
framework if you want to do some serious business without
worrying too much about little details that won't help you
at all to make your deliveries in time. It's opinionated
and I think it's good -- except when I don't want to do
anything "serious".

I think it's too difficult to get a Django project up and
running. You have to deal with your "settings.py",
configure where your templates are, enable static files
serving, write your models upfront, create migrations,
migrate, write the views, then write the templates then
run the server and finally see nothing really happening
because all you did was simply the skeleton of your
application and **now** you can start really making the
things happen as you wanted at first.

What was that, again?

I don't have the patience to deal with so much work only
to put into life something very simple I want to **play**
with.

**I am too lazy!**

## Django REST -ANYTHING- make everything even worse!

Add to all that the decision to have a REST API. Now I have
to write **serializers** or specialized views, too.

**Oh, no!**

I can't do that anymore. Really. Maybe when I was 20 years
old, but not on my age...

## Save your laziness on a DBMS

So, I decided to create a project using Bottle. But, where
should I save my data? Planning to run the app on Heroku, I
couldn't save things on the filesystem. So I found the
"dataset" project and loved it. That was **exactly** what I
needed.

## Models are still nice

I like to design models. What I don't like is to write the
models as code. Yes, you heard it. There should be a better
way.
Since "dataset" is very loose about data and I wanted some
way of validating what was coming from the REST API, I
found it suitable to validate data using "cerberus". So I
needed some place to save the validation schemata, and
decided to use **YAML** for that, since JSON is too verbose
and I am too lazy to comply with JSON strict rules (and
type all that innumerous '{' and '}' and ','...).

Also, model defining should be very simple. Your project
must have a "models.yaml" file on the root directory, from
where models definition, some constraints and validation
rules are going to be loaded.

And that is it.

## Example project using lazywf

There is an example project included on this repository.
You can copy it to use in your own project. It's as simple
as it seems. ;-)


CHANGES.rst
Changelog
---------

0.2.0
~~~~~
* Get HOST, PORT and DEBUG from environment variables.
* Add a __main__ handler, so users don't need to write any code to run.


0.1.2
~~~~~
* Treat the lack of JSON payload properly.


0.1.1
~~~~~

* Does not save "raw" datetimes into database.
* Allow to replace the REST manager.


0.1.0
~~~~~

* First release


tests/conftest.py
import pytest


from lazywf import TheLaziestWebFrameworkEVER


@pytest.fixture
def server():
    class MyTestServer(TheLaziestWebFrameworkEVER):
        def load_models(self):
            self.models = {}

        def db_connect(self):
            self.DATABASE_URL = 'sqlite:///:memory:'
            super().db_connect()

    return MyTestServer()


tests/test_models.py
def test_models_definition(server):
    # TODO: implement tests
    assert 1 == 1


lazywf/rest.py
import datetime

import bottle


def json_operation(f):
    def new_function(self, model_name, *args, **kwargs):
        model = self.models[model_name]
        data = bottle.request.json
        return f(self, model_name, model, data, *args, **kwargs)

    return new_function


class RestOperationsManager:
    def __init__(self, lazy_server):
        self.lazy_server = lazy_server
        self.models = lazy_server.models
        self.db = lazy_server.db

        self.create_routes()

    def create_routes(self):
        self.lazy_server.add_route('/api/<model_name>/', 'POST', callback=self.create)
        self.lazy_server.add_route('/api/<model_name>/<key>', 'PATCH', callback=self.update)
        self.lazy_server.add_route('/api/<model_name>/<key>', 'GET', callback=self.retrieve)
        self.lazy_server.add_route('/api/<model_name>/<key>', 'DELETE', callback=self.delete)
        self.lazy_server.add_route('/api/<model_name>/', 'GET', callback=self.list)

    def validate_model(self, *args, **kwargs):
        return self.lazy_server.validate_model(*args, **kwargs)

    @staticmethod
    def get_parameters_for_search(model, keys):
        key_fields = model['constraints']['keys']
        search_by = {}
        key_index = 0
        for field_name in key_fields:
            search_by[field_name] = keys[key_index]
            key_index += 1

        return search_by

    @json_operation
    def create(self, model_name, model, data):
        is_valid, errors = self.validate_model(model_name, data)

        if not is_valid:
            return bottle.HTTPResponse(status=400, body=errors)

        table = self.db[model_name]

        keys = model['constraints'].get('keys', None)
        if keys:
            search_by = self.get_parameters_for_search(model, [value for key, value in data.items() if key in keys])
            if table.find_one(**search_by):
                return bottle.HTTPResponse(status=400, body={'constraints': 'These fields must be unique: {}'.format(keys)})

        data['created_at'] = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        data['updated_at'] = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')

        table.insert(data)
        return bottle.HTTPResponse(status=201, body=data)

    @staticmethod
    def update_table(model, table, data):
        key_fields = model['constraints']['keys']
        return table.update(data, key_fields)

    @json_operation
    def update(self, model_name, model, data, key):
        keys = key.split(',')
        is_valid, errors = self.validate_model(model_name, data, update=True)

        if not is_valid:
            return bottle.HTTPResponse(status=400, body=errors)

        table = self.db[model_name]

        search_by = self.get_parameters_for_search(model, keys)
        entry = table.find_one(**search_by)
        if not entry:
            return bottle.HTTPResponse(status=404, body={'constraints': 'Not found: {}'.format(search_by)})

        data.update(search_by)
        data['updated_at'] = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')

        self.update_table(model, table, data)
        entry = table.find_one(**search_by)  # XXX: is there a "refresh_from_db", maybe?
        entry['updated_at'] = str(entry['updated_at'])
        return bottle.HTTPResponse(status=200, body=entry)

    def retrieve(self, model_name, key):
        keys = key.split(',')
        table = self.db[model_name]

        model = self.models[model_name]
        search_by = self.get_parameters_for_search(model, keys)
        entry = table.find_one(**search_by)
        if not entry:
            return bottle.HTTPResponse(status=404, body={'constraints': 'Not found: {}'.format(search_by)})

        return bottle.HTTPResponse(status=200, body=entry)

    def delete(self, model_name, key):
        keys = key.split(',')
        table = self.db[model_name]

        model = self.models[model_name]
        search_by = self.get_parameters_for_search(model, keys)
        entry = table.find_one(**search_by)
        if not entry:
            return bottle.HTTPResponse(status=404, body={'constraints': 'Not found: {}'.format(search_by)})

        table.delete(**search_by)
        return bottle.HTTPResponse(status=200, body=entry)

    def list(self, model_name):
        table = self.db[model_name]
        results = {
            'results': [x for x in table],
        }
        return bottle.HTTPResponse(status=200, body=results)


lazywf/__init__.py
import yaml
from os import environ

import bottle
from cerberus import Validator
import dataset
from clint.textui import indent

from .logger import Logger
from .rest import RestOperationsManager


class TheLaziestWebFrameworkEVER:
    def __init__(self):
        self.DATABASE_URL = environ.get('DATABASE_URL', None)

        self.logger = Logger(self)

        # Bottle app:
        self.app = bottle.Bottle()

        self.load_models()
        self.db_connect()

        self.create_internal_routes()
        self.create_routes()
        self.load_rest_manager()

        self.enable_cors()

    def load_rest_manager(self):
        if self.models:
            self.rest_manager = RestOperationsManager(self)

    def add_cors_headers(self, response=None):
        response = response or bottle.response

        if self.cors_origins is None:
            return response

        the_headers = response.headers

        the_headers['Access-Control-Allow-Origin'] = self.cors_origins
        the_headers['Access-Control-Allow-Methods'] = 'PUT, GET, POST, DELETE, OPTIONS'
        the_headers['Access-Control-Allow-Headers'] = 'Origin, Accept, Content-Type, X-Requested-With, X-CSRF-Token'

    def enable_cors(self):
        self.cors_origins = environ.get('CORS_ORIGINS', None)
        self.app.hook('after_request')(self.add_cors_headers)

    # ---------------------------------------
    # Models and Database:
    def load_models(self):
        try:
            return self.do_load_models()
        except FileNotFoundError:
            self.logger.warning('NOT loading models: models.yaml file does not exist.')
            self.models = None
            return

    def do_load_models(self):
        with open('models.yaml') as models_file:
            self.models = yaml.load(models_file)

    def db_connect(self):
        if self.DATABASE_URL and self.models:
            self.logger.info('Connecting to database')
            with indent(4):
                self.db = dataset.connect(self.DATABASE_URL)
                self.logger.info('Database URL: {}'.format(self.db.url))
        else:
            self.logger.warning('NOT connecting to database: DATABASE_URL environment variable is not set or no models definitions available.')
            self.db = None

    def get_table(self, table_name):
        return self.db[table_name]

    # -------------------
    def add_route(self, path, method, callback, **kwargs):
        self.logger.info('add_route: {} {}'.format(method, path))
        return self.app.route(path, method, callback=callback, **kwargs)

    def create_routes(self):
        # Child class can implement.
        pass

    def create_internal_routes(self):
        self.add_route('/static/<filename:path>', 'GET', callback=self.serve_static)

    def validate_model(self, name, data, update=False):
        if not data:
            return False, {"__all__": "No document provided."}

        schema = self.models[name]['validations']
        v = Validator(schema)
        return v.validate(data, update=update), v.errors

    # Error page:
    @bottle.error(404)  # TODO: check if this works properly.
    def error404(self, error):
        return 'Not found ({}).'.format(error)

    # Static files:
    def serve_static(self, filename):
        response = bottle.static_file(filename, root='static')
        self.add_cors_headers(response)
        return response

    # REST API:
    # Templates
    def render(self, template_name, context={}):
        return bottle.template('templates/{}'.format(template_name), **context)

    # Run the server!!!
    def run(self):
        port = int(environ.get('PORT', 8080))
        host = environ.get('HOST', 'localhost')
        debug_mode = environ.get('DEBUG', 'True') == 'True'
        return bottle.run(self.app, host=host, port=port, debug_mode=debug_mode)
